aten/src/ATen/native/cuda/Math.cuh

```diff
@@ -0,0 +1,91 @@
+#pragma once
+#include <type_traits>
+
+namespace at {
+namespace native {
+
+/*
+* The following function was converted to CUDA form from code that comes
+* with the following copyright notice. It has been released under the BSD license.
+*
+* Cephes Math Library Release 2.8:  June, 2000
+* Copyright 1984, 1987, 1992, 2000 by Stephen L. Moshier
+*/
+template <typename T, typename accreal>
+static inline __host__ __device__ T calc_digamma(T in) {
+  using compute_type = typename std::conditional<std::is_same<T, at::Half>::value, accreal, T>::type;
+  static const double PI_f64 = 3.14159265358979323846;
+  compute_type PSI_10 = 2.25175258906672110764;
+  compute_type A[] = {
+      8.33333333333333333333E-2,
+      -2.10927960927960927961E-2,
+      7.57575757575757575758E-3,
+      -4.16666666666666666667E-3,
+      3.96825396825396825397E-3,
+      -8.33333333333333333333E-3,
+      8.33333333333333333333E-2,
+  };
+
+  auto x = static_cast<compute_type>(in);
+  if (x == 0) {
+    return static_cast<T>(INFINITY);
+  }
+
+  bool x_is_integer = x == ::floor(x);
+  compute_type result = 0;
+  if (x < 0) {
+    if (x_is_integer) {
+      return static_cast<T>(INFINITY);
+    }
+    // Rounding errors in tan's input can really affect the output
+    // for extreme values, so we always perform this computation in double.
+    result = static_cast<compute_type>(- PI_f64 / ::tan(PI_f64 * static_cast<double>(x)));
+    x = 1 - x;
+  }
+
+  while (x < 10) {
+    result -= 1 / x;
+    x += 1;
+  }
+  if (x == 10) {
+    return static_cast<T>(result + PSI_10);
+  }
+
+  compute_type y = 0;
+  if (x < 1.0e17) {
+    compute_type z = 1.0 / (x * x);
+
+    compute_type polevl_result = 0;
+    for (int i = 0; i <= 6; i++) {
+      polevl_result = polevl_result * z + A[i];
+    }
+    y = z * polevl_result;
+  }
+
+  return static_cast<T>(::log(x) - (0.5 / x) - y + result);
+}
+
+template <typename T, typename accreal>
+static inline __host__ __device__ T calc_trigamma(T in) {
+  using compute_type = typename std::conditional<std::is_same<T, at::Half>::value, accreal, T>::type;
```

Comment:
Do we need to make `accreal` a parameter? Since `accreal` should be always `float` for `Half`, the following template mapping might help reduce code redundancy for now and future (untested):

```c++
template <typename T>
struct accreal {
    using type = typename T;
};

template <>
struct accreal<at::Half> {
    using type = float;
};
```

To use, simply call `using compute_type = typename accreal<my_type>::type`

Reply:
Removed accreal and left scalar_t that will be float for Half